### PR TITLE
CI: Run dev tests on Scientific Python nightly wheels

### PR DESCRIPTION
**Pull request by weiji14**
**Description of proposed changes**

Check that PyGMT will work on future versions of numpy>=2.0.0, pandas>=2.1.0, xarray>=2023.7.1 and so on.

References:
- Pre-built wheels are at https://anaconda.org/scientific-python-nightly-wheels
- [SPEC](https://scientific-python.org/specs) 4 — Using and Creating Nightly Wheels at https://scientific-python.org/specs/spec-0004

Extends #1105 when `pip install --pre` started being used. The nightly wheels from https://pypi.anaconda.org/scientific-python-nightly-wheels/simple are a lot more up to date than those on the default PyPI index.

**Reminders**

- [ ] Run `make format` and `make check` to make sure the code follows the style guide.
- [ ] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [ ] Write detailed docstrings for all functions/methods.
- [ ] If wrapping a new module, open a 'Wrap new GMT module' issue and submit reasonably-sized PRs.
- [ ] If adding new functionality, add an example to docstrings or tutorials.
- [ ] Use underscores (not hyphens) in names of Python files and directories.

**Slash Commands**

You can write slash commands (`/command`) in the first line of a comment to perform
specific operations. Supported slash commands are:

- `/format`: automatically format and lint the code
- `/test-gmt-dev`: run full tests on the latest GMT development version

### CODE REVIEW

File: .github/workflows/ci_tests_dev.yaml [hunk 1]
```diff
@@ -17,7 +18,7 @@ on:
   # push:
   #   branches: [ main ]
   pull_request:
-    types: [ready_for_review]
+    types: [opened, reopened, synchronize, ready_for_review]
```
Comment (weiji14) [inline .github/workflows/ci_tests_dev.yaml:21]: TODO revert this before merging.

```suggestion
    types: [ready_for_review]
```

**Comment by weiji14** [comment]
Nice, no visible deprecation warnings or errors with using the nightly versions of NumPy, Pandas, Xarray, etc :tada:. See https://github.com/GenericMappingTools/pygmt/actions/runs/5594537134/jobs/10229490969?pr=2612#step:16:723

```
=================================== FAILURES ===================================
__________________ test_earth_relief_03s_default_registration __________________

    def test_earth_relief_03s_default_registration():
        """
        Test that the grid returned by default for the 3 arc-second resolution has
        a "gridline" registration.
        """
        data = load_earth_relief(resolution="03s", region=[-10, -9.8, 4.9, 5])
        assert data.shape == (121, 241)
        assert data.gmt.registration == 0
        npt.assert_allclose(data.coords["lat"].data.min(), 4.9)
        npt.assert_allclose(data.coords["lat"].data.max(), 5)
        npt.assert_allclose(data.coords["lon"].data.min(), -10)
        npt.assert_allclose(data.coords["lon"].data.max(), -9.8)
>       npt.assert_allclose(data.min(), -2069.996)

../pygmt/tests/test_datasets_earth_relief.py:263: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

args = (<function assert_allclose.<locals>.compare at 0x7f7acb461a80>, array(-2131.9707, dtype=float32), array(-2069.996))
kwds = {'equal_nan': True, 'err_msg': '', 'header': 'Not equal to tolerance rtol=1e-07, atol=0', 'verbose': True}

    @wraps(func)
    def inner(*args, **kwds):
        with self._recreate_cm():
>           return func(*args, **kwds)
E           AssertionError: 
E           Not equal to tolerance rtol=1e-07, atol=0
E           
E           Mismatched elements: 1 / 1 (100%)
E           Max absolute difference: 61.97470312
E           Max relative difference: 0.02993953
E            x: array(-2131.9707, dtype=float32)
E            y: array(-2069.996)

../../../../micromamba/envs/pygmt/lib/python3.11/contextlib.py:81: AssertionError
=============================== warnings summary ===============================
pygmt/helpers/decorators.py::pygmt.helpers.decorators.use_alias
  <doctest pygmt.helpers.decorators.use_alias[1]>:1: SyntaxWarning: Short-form parameter (R) is not recommended. Use long-form parameter 'region' instead.

pygmt/helpers/decorators.py::pygmt.helpers.decorators.use_alias
  <doctest pygmt.helpers.decorators.use_alias[1]>:1: SyntaxWarning: Short-form parameter (J) is not recommended. Use long-form parameter 'projection' instead.

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
==================================== PASSES ====================================
...
```

Only error is on `test_earth_relief_03s_default_registration` which is failing on the main branch too and reported at https://github.com/GenericMappingTools/pygmt/issues/2511#issue-1679305693

File: .github/workflows/ci_tests_dev.yaml (hunk 1 above)
Comment (weiji14) [reply .github/workflows/ci_tests_dev.yaml]: Reverted in 8bda1a78d237e37f2b0b28873eda1e7161ec6149

**Review by michaelgrund — APPROVED** [review]

File: .github/workflows/ci_tests_dev.yaml [hunk 2]
```diff
@@ -1,6 +1,7 @@
 # Tests PyGMT with GMT dev version on Linux/macOS/Windows
```
Comment (yvonnefroehlich) [inline .github/workflows/ci_tests_dev.yaml:1]: To be consistent with the rest of the documentation and tests (see PR #2578). Similar changes are needed in the files `ci_tests.yaml` and `ci_tests_legacy.yaml`, please.

```suggestion
# Test PyGMT with GMT dev version on Linux/macOS/Windows
```

Comment (weiji14) [reply .github/workflows/ci_tests_dev.yaml:1]: Cool, let me do that tomorrow when I'm back at my computer. Hard to edit multiple files on mobile 😬

Comment (yvonnefroehlich) [reply .github/workflows/ci_tests_dev.yaml:1]: No hurry, that's only a small change 🙂.

**Review by yvonnefroehlich — APPROVED** [review]

File: .github/workflows/ci_tests_dev.yaml (hunk 2 above)
Comment (weiji14) [reply .github/workflows/ci_tests_dev.yaml]: Ok, done in 8b203691294e44a4e9ebc5318c06fcacd3e1845b

**Pull request merged**